client/model.go

```diff
@@ -50,6 +54,100 @@ func (client *Client) Remodel(b []byte) (changeID string, err error) {
 	return client.doAsync("POST", "/v2/model", nil, headers, bytes.NewReader(data))
 }
 
+// RemodelOffline tries to remodel the system with the given model assertion
+// and local snaps and assertion files.
+func (client *Client) RemodelOffline(
+	model []byte, snapPaths, assertPaths []string) (changeID string, err error) {
+
+	// Check if all files exist before starting the go routine
+	snapFiles, err := checkAndOpenFiles(snapPaths)
+	if err != nil {
+		return "", err
+	}
+	assertsFiles, err := checkAndOpenFiles(assertPaths)
+	if err != nil {
+		return "", err
+	}
+
+	pr, pw := io.Pipe()
+	mw := multipart.NewWriter(pw)
+	go sendRemodelFiles(model, snapPaths, snapFiles, assertsFiles, pw, mw)
+
+	headers := map[string]string{
+		"Content-Type": mw.FormDataContentType(),
+	}
+
+	_, changeID, err = client.doAsyncFull("POST", "/v2/model", nil, headers, pr, doNoTimeoutAndRetry)
+	return changeID, err
+}
+
+func checkAndOpenFiles(paths []string) ([]*os.File, error) {
+	var files []*os.File
+	for _, path := range paths {
+		f, err := os.Open(path)
+		if err != nil {
+			for _, openFile := range files {
+				openFile.Close()
+			}
+			return nil, fmt.Errorf("cannot open %q: %w", path, err)
+		}
+
+		files = append(files, f)
+	}
+
+	return files, nil
+}
+
+func sendRemodelFiles(model []byte, paths []string, files, assertFiles []*os.File, pw *io.PipeWriter, mw *multipart.Writer) {
+	defer func() {
+		for _, f := range files {
+			f.Close()
+		}
+	}()
+
+	if err := mw.WriteField("new-model", string(model)); err != nil {
+		pw.CloseWithError(err)
+		return
+	}
+
+	for _, file := range assertFiles {
+		if err := sendPartFromFile(file,
+			func() (io.Writer, error) {
+				return mw.CreateFormField("assertion")
```

Comment:
I have slightly preference if we used CreatFormPart here and gave the assertion parts proper Content-Types

Reply:
Done